File: 백준/Silver/18405.경쟁적 전염/경쟁적 전염.cc
#include <bits/stdc++.h>

//멀티 쓰레드 느낌

using namespace std;

int n, k, cnt, s, x, y;
bool check[222][222];
int board[222][222];
int dx[4] = {0, 1, 0, -1};
int dy[4] = {-1, 0, 1, 0};
int vsize;

queue<pair<int, int>> q;
queue<pair<int, int>> temp_q;

void bfs(int sec)
{
    while (sec != s)
    {
        while (!q.empty())
        {
            auto &top = q.front();
            int cx = top.second;
            int cy = top.first;
            //cout << cy << ' ' << cx << '\n';

            q.pop();

            check[cy][cx] = true;

            for (int k = 0; k < 4; k++)
            {
                int nx = cx + dx[k];
                int ny = cy + dy[k];

                if (0 < nx && nx <= n && 0 < ny && ny <= n)
                {
                    if (!check[ny][nx] && board[ny][nx] == 0)
                    {
                        check[ny][nx] = true;
                        board[ny][nx] = board[cy][cx];
                        temp_q.push({ny, nx});
                    }
                }
            }
        }
        sec++;

        q = temp_q;
        while (!temp_q.empty())
        {
            temp_q.pop();
        }
    }
    cout << board[y][x] << '\n';
}

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    cin >> n >> k;

    for (int i = 1; i <= n; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            cin >> board[i][j];
        }
    }

    for (int l = 1; l <= k; l++)
    {
        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= n; j++)
            {
                if (board[i][j] == l)
                    q.push({i, j});
            }
        }
    }

    cin >> s >> y >> x;

    bfs(0);

    return 0;
}

File: 백준/Gold/18405. 경쟁적 전염/경쟁적 전염.kt
import java.io.*
import java.util.*

val scan = FastReader()
val sb = StringBuilder()

var N = 0
var K = 0
var S = 0
var X = 0
var Y = 0

lateinit var board: Array<IntArray>
lateinit var check: Array<BooleanArray>

var q: Queue<Pair<Int,Int>> = LinkedList()
var temp_q: Queue<Pair<Int,Int>> = LinkedList()

val dx = intArrayOf(0, 1, 0, -1)
val dy = intArrayOf(-1, 0, 1, 0)

fun main() {
    input()
    pro()
}

fun input() {
    N = scan.nextInt()
    K = scan.nextInt()

    board = Array(N + 1) { IntArray(N + 1) }
    check = Array(N + 1) { BooleanArray(N + 1) }

    for (i in 1..N) {
        for (j in 1..N) {
            board[i][j] = scan.nextInt()
        }
    }

    //virus 넘버링
    for (v in 1..K) {
        for (i in 1..N) {
            for (j in 1..N) {
                if (board[i][j] == v)
                    q.offer(Pair(i, j))
            }
        }
    }

    S = scan.nextInt()
    X = scan.nextInt()
    Y = scan.nextInt()
}

fun pro() {
    bfs(0)
}

fun bfs(sec: Int) {
    var second = sec
    while (second != S) {
        while (!q.isEmpty()) {
            val top = q.poll()
            val cx = top.first
            val cy = top.second

            check[cx][cy] = true

            for (k in 0 until 4) {
                val nx = cx + dx[k]
                val ny = cy + dy[k]

                if (nx in 1..N && ny in 1..N) {
                    if (!check[nx][ny] && board[nx][ny] == 0) {
                        check[nx][ny] = true
                        board[nx][ny] = board[cx][cy]
                        //show()
                        temp_q.offer(Pair(nx, ny))
                    }
                }
            }
        }
        second += 1

        while (!temp_q.isEmpty()) {
            val element = temp_q.poll()
            q.offer(element)
        }
    }

    sb.append(board[X][Y]).append("\n")
    print(sb.toString())
}

fun show() {
    for(i in 1..N) {
        for (j in 1..N) {
            print("${board[i][j]} ")
        }
        println()
    }
    println()
}

class FastReader {
    var br: BufferedReader
    var st: StringTokenizer? = null

    constructor() {
        br = BufferedReader(InputStreamReader(System.`in`))
    }

    constructor(s: String?) {
        br = BufferedReader(FileReader(File(s)))
    }

    operator fun next(): String {
        while (st == null || !st!!.hasMoreElements()) {
            try {
                st = StringTokenizer(br.readLine())
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        return st!!.nextToken()
    }

    fun nextInt(): Int {
        return next().toInt()
    }

    fun nextLong(): Long {
        return next().toLong()
    }

    fun nextDouble(): Double {
        return next().toDouble()
    }

    fun nextLine(): String {
        var str = ""
        try {
            str = br.readLine()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return str
    }
}